rfcs/detecting_sensitive_textual_content/20230309-project_proposal_detecting_sensitive_textual_content.md

```diff
@@ -0,0 +1,379 @@
+
+## Reviewers
+
+- [ ] @aetherunbound
+- [ ] @krysal
+
+## Project summary
+
+Detect results with sensitive textual content by checking their textual content
+against a list of sensitive terms. Once results with sensitive terms in their
+textual content are determined, blur those results on the frontend.
+
+## Goals
+
+This advances Openverse's goal of increasing the safety of the platform.
+
+## Requirements
+
+> **A note on tone**: This section should be mostly readable and understandable
+> by casual, non-technical readers. However, certain aspects of it unavoidably
+> delve into technical details in order to describe how features fit together or
+> considerations that will need to be kept in mind during implementation
+> planning. These need to be documented _somewhere_, and displacing them from
+> the requirements felt strange. If you are not interested in these particular
+> technical details, please skip the two sections titled "Implementation
+> details" and the section on "How to blur results". These sections in
+> particular consider technical issues of the project that may confound readers
+> unfamiliar with software development whether applied to Openverse or more
+> generally.
+
+### API
+
+#### Filtering
+
+The Django API is able to filter out results that include sensitive terms in
+their textual content. It does so without degrading search performance.
+
+Filtering of results with sensitive terms in their textual content occurs when
+`include_sensitive_results=True`. This parameter will supplant the `mature=True`
+as a more comprehensively and descriptively named parameter. `mature=True`
+should still work, but it should just be an alias for
+`include_sensitive_results` and should be marked as deprecated in or altogether
+excluded from the API documentation.
+
+##### Implementation details
+
+This will likely be implemented as a secondary index of results that do not
+include the sensitive terms, as explored in
+https://github.com/WordPress/openverse-api/pull/1108.
+
+#### Designation of results with sensitive terms
+
+Results that include sensitive terms in their textual content are so denoted in
+the API results in a new result field named `sensitivity`. Sensitivity should be
+an array that should be empty for results where `mature = false` and the textual
+content does not contain any sensitive terms. For results marked as `mature`,
+the array should include a `"mature"` string. For results with sensitive terms
+in their textual content, the array should include a `"sensitive_text"` string.
+At this time we will not be denoting _what_ sensitive terms are present nor will
+we be developing any categories of terms. This approach leaves open the
+potential for that in the future, however, by allowing the array to include
+specific tokens for particular categories of sensitivity, whether derived from a
+categorisation of the sensitive terms list, or from an image classification API.
```

Comment:
Does "token" here mean to some reference to a sensitivity category, like the id of a row in a `sensitivity_categories` table, for example? I just want to make sure there isn't anything else happening here I wasn't understanding.

Reply:
It's intentionally non-specific. It could be a string token literal (`sensitive_terms::violence`) or an obfuscated ID if we felt that was reasonable/necessary. In either case it's a constant that a client (our frontend or others) can use to display additional information about the result or even ignore it completely.

However, those would all be future ideas. My goal here was to avoid recommending us use a new boolean property (or overloading the `mature` property) as that boolean would become quickly obsolete if we expanded the tokenised reasons list and started using it for display information.

Reply:
That makes perfect sense, thanks.